cfg/matchrule/matchrule.go

```diff
@@ -0,0 +1,232 @@
+package matchrule
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"strings"
+)
+
+type Mode byte
+
+var (
+	_ json.Unmarshaler = (*Mode)(nil)
+)
+
+func (m *Mode) UnmarshalJSON(i []byte) error {
+	switch {
+	case bytes.Equal(i, modePrefixBytes):
+		*m = ModePrefix
+	case bytes.Equal(i, modeContainsBytes):
+		*m = ModeContains
+	case bytes.Equal(i, modeSuffixBytes):
+		*m = ModeSuffix
+	default:
+		return fmt.Errorf("unknown mode %s", string(i))
+	}
+	return nil
+}
+
+const (
+	ModePrefix Mode = iota
+	ModeContains
+	ModeSuffix
+)
+
+var (
+	modePrefixBytes   = []byte(`"prefix"`)
+	modeContainsBytes = []byte(`"contains"`)
+	modeSuffixBytes   = []byte(`"suffix"`)
+)
+
+type Rule struct {
+	// > @3@4@5@6
+	// >
+	// > Values to match at least one value.
+	Values []string `json:"values" default:"" slice:"true"` // *
+
+	// > @3@4@5@6
+	// >
+	// > Matching mode. Allowed modes: `prefix`, `contains`, `suffix`. Defaults to `contains`.
+	Mode Mode `json:"mode"` // *
+
+	// > @3@4@5@6
+	// >
+	// > CaseInsensitive is the truth then Match results in a lowercase search value.
+	// > Not available fo the ModeContains because performance issues may arise.
+	CaseInsensitive bool `json:"case_insensitive"` // *
+
+	// > @3@4@5@6
+	// >
+	// > Flag indicating whether to negate the result of matching rule.
+	Invert bool `json:"invert"` // *
+
+	minValueSize int
+	maxValueSize int
+}
+
+func NewRule(values []string, mode Mode, caseInsensitive bool) Rule {
+	if caseInsensitive && mode == ModeContains {
+		panic(fmt.Errorf("сase insensitive for the 'contains' condition is not supported"))
+	}
+
+	for i := range values {
+		if caseInsensitive {
+			values[i] = strings.ToLower(values[i])
+		}
+	}
+
+	return Rule{
+		Values:          values,
+		Mode:            mode,
+		CaseInsensitive: caseInsensitive,
+	}
+}
+
+func (e *Rule) calcValueSizes() {
+	minValueSize := math.MaxInt
+	maxValueSize := 0
+	for i := range e.Values {
+		if len(e.Values[i]) > maxValueSize {
+			maxValueSize = len(e.Values[i])
+		}
+		if len(e.Values[i]) < minValueSize {
+			minValueSize = len(e.Values[i])
+		}
+	}
+	e.minValueSize = minValueSize
+	e.maxValueSize = maxValueSize
+}
+
+func (e *Rule) Match(raw []byte) bool {
+	ok := e.match(raw)
+	if e.Invert {
+		ok = !ok
+	}
+	return ok
+}
+
+func (e *Rule) match(raw []byte) bool {
+	if e.maxValueSize == 0 {
+		e.calcValueSizes()
```

Comment:
let's calc the min/max size in the constructor

Reply:
Added lowering values for case insensitive mode and renamed method to `processValues`